pkg/syncer/util.go

```diff
@@ -623,3 +631,160 @@ func getPVsInBoundAvailableOrReleasedForVc(ctx context.Context, metadataSyncer *
+	}
+
+	containerCluster := cnsvsphere.GetContainerCluster(clusterIDforVolumeMetadata,
+		vcHostObj.User, metadataSyncer.clusterFlavor,
+		metadataSyncer.configInfo.Cfg.Global.ClusterDistribution)
+
+	createSpec := &cnstypes.CnsVolumeCreateSpec{
+		Name:       pv.Name,
+		VolumeType: volumeType,
+		Metadata: cnstypes.CnsVolumeMetadata{
+			ContainerCluster:      containerCluster,
+			ContainerClusterArray: []cnstypes.CnsContainerCluster{containerCluster},
+			EntityMetadata:        metadataList,
+		},
+	}
+
+	if volumeType == common.BlockVolumeType {
+		createSpec.BackingObjectDetails = &cnstypes.CnsBlockBackingDetails{
+			CnsBackingObjectDetails: cnstypes.CnsBackingObjectDetails{},
+			BackingDiskId:           pv.Spec.CSI.VolumeHandle,
+		}
+	} else {
+		createSpec.BackingObjectDetails = &cnstypes.CnsVsanFileShareBackingDetails{
+			CnsFileBackingDetails: cnstypes.CnsFileBackingDetails{
+				BackingFileId: pv.Spec.CSI.VolumeHandle,
+			},
+		}
+	}
+	log.Debugf("PVUpdated: vSphere CSI Driver is creating volume %q with create spec %+v",
+		pv.Name, spew.Sdump(createSpec))
+	_, _, err := cnsVolumeMgr.CreateVolume(ctx, createSpec)
+	if err != nil {
+		log.Errorf("PVUpdated: Failed to create disk %s with error %+v", pv.Name, err)
+		return err
+	} else {
+		log.Infof("PVUpdated: vSphere CSI Driver has successfully marked volume: %q as the container volume.",
+			pv.Spec.CSI.VolumeHandle)
+
+		if isMultiVCenterFssEnabled && len(metadataSyncer.configInfo.Cfg.VirtualCenter) > 1 {
+			// Create CNSVolumeInfo CR for the volume ID.
+			err = volumeInfoService.CreateVolumeInfo(ctx, pv.Spec.CSI.VolumeHandle, vcHost)
+			if err != nil {
+				log.Errorf("PVUpdated: Failed to store volumeID %q for vCenter %q in CNSVolumeInfo CR. Error: %+v",
+					pv.Spec.CSI.VolumeHandle, vcHost, err)
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func createMissingFileVolumeInfoCrs(ctx context.Context, metadataSyncer *metadataSyncInformer) {
+
+	log := logger.GetLogger(ctx)
+
+	// Get all K8s volumes in "Bound", "Available" or "Released" states.
+	allPvs, err := getPVsInBoundAvailableOrReleased(ctx, metadataSyncer)
+	if err != nil {
+		log.Errorf("Failed to get PVs in the cluster. Err: %v", err)
+		return
+	}
+
+	fileVolumes := make([]*v1.PersistentVolume, 0)
+	for _, pv := range allPvs {
+		if pv.Spec.CSI == nil {
+			if metadataSyncer.coCommonInterface.IsFSSEnabled(ctx, common.CSIMigration) &&
+				pv.Spec.VsphereVolume != nil {
+				log.Errorf("In-tree volumes are not supported on a multi VC set up."+
+					"Found in-tree volume %s.", pv.Name)
+				return
+			}
+			log.Errorf("Invalid PV %s with empty volume handle.", pv.Name)
+			return
+		}
+		// Check if the PV is a file volume.
+		if IsMultiAttachAllowed(pv) {
+			fileVolumes = append(fileVolumes, pv)
+		}
+	}
+
+	if len(fileVolumes) == 0 {
+		log.Debugf("There are no file volumes on the cluster")
+		return
+	}
+
+	fileVolumesWithMissingCrs := make([]*v1.PersistentVolume, 0)
```

Comment:
Do you see any race between static file volume creation and this full sync logic?

Reply:
I do not really see a problem here.

Case 1

Metadatasyncer creates PV and is in process of creating CR but before that full sync kicks in and notices that CR is missing and attempts to create the volume and hence the CR. 
It does not matter whether metadatasyncer of full sync creates the CR first.

Case 2

Before metadatasyncer can create the volume/CR, full sync creates it first. Again, it does not matter which of them is able to create it first.

CNSVolumeInfoService takes care of already exists https://github.com/kubernetes-sigs/vsphere-csi-driver/blob/master/pkg/internalapis/cnsvolumeinfo/cnsvolumeinfoservice.go#L175

CNS should also ideally take care of volume already exists: https://github.com/kubernetes-sigs/vsphere-csi-driver/blob/master/pkg/common/cns-lib/volume/util.go#L287

However, CNS is sending back `CnsAlreadyRegisteredFault` only for block volumes. I have raised a bug against CNS to fix this. Bug #3297537